### PR TITLE
Initial work to check that a db is up to a certain version before running queries

### PR DESCRIPTION
**Pull request by davidegrohmann**
This behaviour is not used yet since there is no support yet in the drivers.

### CODE REVIEW

File: community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTrackingTest.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            fail( "should have thrown" );
+        }
+        catch ( TransactionFailureException e )
+        {
+            // then all good!
+        }
+    }
+
+    @Test( timeout = 500 )
+    public void shouldNotUpdateVersionIfTheInitialVersionIsLessThanZero() throws Exception
```
Comment (jimwebber) [inline community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTrackingTest.java]: Could the name include the why (no transactions yet applied) rather than the how (negative tx id)?

File: pom.xml [hunk 2]
```diff
@@ -632,7 +632,7 @@
       <dependency>
         <groupId>io.netty</groupId>
         <artifactId>netty-all</artifactId>
-        <version>4.0.28.Final</version>
+        <version>4.0.37.Final</version>
```
Comment (davidegrohmann) [inline pom.xml]: sneaky! :)

Maybe a separate commit?

File: community/bolt/src/main/java/org/neo4j/bolt/BoltKernelExtension.java [hunk 3]
```diff
@@ -133,61 +141,58 @@ public Lifecycle newInstance( KernelContext context, Dependencies dependencies )
...
+        StandardSessions standardSessions = life.add(
+                new StandardSessions( api, dependencies.usageData(), logService, dependencies.txBridge(),
+                        authentication, dependencies.neoStoreDataSource() ) );
+        ThreadedSessions threadedSessions = new ThreadedSessions( standardSessions, scheduler, logService );
+        Sessions sessions = new MonitoredSessions( dependencies.monitors(), threadedSessions, Clock.systemUTC() );
+
+        List<ProtocolInitializer> connectors =
+                config.view( enumerate( GraphDatabaseSettings.Connector.class ) ).map( BoltConnector::new ).filter(
+                        ( connConfig ) -> BOLT.equals( config.get( connConfig.type ) ) &&
+                                config.get( connConfig.enabled ) ).map( ( connConfig ) -> {
```
Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/BoltKernelExtension.java]: Since this code is already touched by formatting, please pull this big lambda into a method. This whole stream processing is quite hard to read otherwise.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java [hunk 4]
```diff
@@ -674,21 +694,35 @@ State error( SessionStateMachine ctx, Neo4jError err )
...
         void bindTransactionToCurrentThread( KernelTransaction tx );
         void unbindTransactionFromCurrentThread();
         RecordStream run( SessionStateMachine ctx, String statement, Map<String, Object> params )
                 throws KernelException;
         AuthenticationResult authenticate( Map<String, Object> authToken ) throws AuthenticationException;
         void udcRegisterClient( String clientName );
         Statement currentStatement();
+        TransactionIdTracker versionTracking( long startingVersion );
+
+        interface TransactionIdTracker
```
Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java]: Please extract this as top level interface. Functionality looks generic enough and full name `SessionStateMachine.SPI.TransactionIdTracker` is not very pleasant.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java [hunk 5]
```diff
@@ -674,21 +694,35 @@ State error( SessionStateMachine ctx, Neo4jError err )
...
+        {
+            void assertUpToDate() throws TransactionFailureException;
```
Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java]: Looks strange that this method throws `TransactionFailureException`. Why assertion should know about transactions?
Maybe make this method return boolean and then use it to throw `TransactionFailureException` in `StandardStateMachineSPI#beginTransaction()`.

Also method name does not describe what it is doing. It should be smth like `waitUntilKnownTxOnParWithLastClosedTx`.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java [hunk 6]
```diff
@@ -674,21 +694,35 @@ State error( SessionStateMachine ctx, Neo4jError err )
...
+
+            void updateVersion( long version );
```
Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java]: This method should not be about versions. Consider `void updateClosedTx(long txId)`.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    public void assertUpToDate() throws TransactionFailureException
+    {
+        if ( txId <= BASE_TX_ID )
+        {
+            return;
+        }
+
+        try
+        {
+            Predicates.await( () -> txId <= transactionIdStore.get().getLastClosedTransactionId(), timeout, timeoutUnit,
```
Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java]: This method will throw `TimeoutException` if condition was not met. So here we basically use exceptions for flow control, which is not nice. Thrown `TimeoutException` will also use condition for it's message. Message is not going to be pretty with this condition. Please extract it in a `Supplier` implementation with overridden `toString()`.

File: community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/SessionAuthIT.java [hunk 8]
```diff
@@ -51,7 +51,7 @@ public void shouldGiveCredentialsExpiredStatusOnExpiredCredentials() throws Thro
         session.init( "TestClient/1.0.0", map(
                 "scheme", "basic",
                 "principal", "neo4j",
-                "credentials", "neo4j" ), null, recorder );
+                "credentials", "neo4j" ), -1, null, recorder );
```
Comment (lutovich) [inline community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/SessionAuthIT.java:54]: Maybe use `TransactionIdStore.BASE_TX_ID` instead of all these `-1` because later is not really a valid tx id.

File: pom.xml (hunk 2 above)
Comment (lutovich) [inline pom.xml]: Is this change relevant to this PR?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java [hunk 9]
```diff
@@ -142,7 +143,7 @@ TransactionWriteState upgradeToSchemaWrites() throws InvalidTransactionTypeKerne
     private TransactionHooks.TransactionHooksState hooksState;
     private final KernelStatement currentStatement;
     private final StorageStatement storageStatement;
-    private CloseListener closeListener;
+    private final List<CloseListener> closeListeners = new ArrayList<>( 2 );
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: As I've said previously, I'm against these listeners because they make code really hard to understand and debug. Especially when there are multiple of them. `SessionStateMachine.State` operates on `KernelTransaction` which is an internal interface. We can change it's `void #close()` to `long #commit()` that returns transaction id. Then it'll be possible to call `commit()` in the state machine and update `TransactionIdTracker` with the returned tx id.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java [hunk 10]
```diff
@@ -621,7 +623,7 @@ private void afterRollback()
     {
         try
         {
-            closeTransaction();
+            closeTransaction( -1 );
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java]: Why is this value `-1`? Does it rely on the implementation of `TransactionIdTracker` that is much higher in the stack?

**Comment by lutovich** [comment]
@davidegrohmann first review round completed. I dislike the chosen design with listeners on `KernelTransaction#close()`. When fixing comments, please add new commits and do not amend old ones - this makes review easier.

File: community/bolt/src/main/java/org/neo4j/bolt/BoltKernelExtension.java (hunk 3 above)
Comment (davidegrohmann) [inline community/bolt/src/main/java/org/neo4j/bolt/BoltKernelExtension.java]: I don't think refactoring unreadable code should be part of this PR.

Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/BoltKernelExtension.java]: Code formatting should also not be part of this PR.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java (hunk 5 above)
Comment (davidegrohmann) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java]: No,I think I rather change the interface name, it is too implementation specific. The concept of "database version" might be actually different  from transaction ids.

I cannot make it return a boolean since the waiting part should be the tracker responsibility, I will throw a different exception.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java (hunk 6 above)
Comment (davidegrohmann) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java]: No as I said in the other comment I rather change the interface name

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java (hunk 7 above)
Comment (davidegrohmann) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java]: I don't understand how you can claim here we use `Exception`s for control flow, do you think it is better  to duplicated the code and write the same `while` loop here instead of using `Predicates.await`?

I'll remove the caught exception as a cuase since it is not relevant and I'll improve the message of the exception we throw.

File: community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/SessionAuthIT.java (hunk 8 above)
Comment (davidegrohmann) [inline community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/SessionAuthIT.java:54]: That's the intent, `BASE_TX_ID` has a different behaviour than `-1`.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 9 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: No, what you suggest will involve too many code changes and a completely different API around `KernelTransaction`. I am not willing to take on such a big refactor.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 10 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java]: `-1` means that we are rollbacking the transaction.

File: pom.xml (hunk 2 above)
Comment (davidegrohmann) [inline pom.xml]: No but it is not harmful either. It leaves in its own commit now.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java (hunk 7 above)
Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java]: This code decides what to do based on if this line throws or now. Isn't it a flow control with exceptions? It is especially visible when you'll ignore `TimeoutException` and would not add it as a cause.
I think there should be a `Predicates.await()` method that returns a boolean.

File: community/bolt/src/main/java/org/neo4j/bolt/BoltKernelExtension.java (hunk 3 above)
Comment (davidegrohmann) [inline community/bolt/src/main/java/org/neo4j/bolt/BoltKernelExtension.java]: I'll revert it then

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 9 above)
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: Big refactoring is a necessary measure here. Otherwise we'll drown in the swamp of listeners. I can do/try such refactoring myself, if it is needed for this feature.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java (hunk 7 above)
Comment (davidegrohmann) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java]: No it simply wraps an exception in another, that's not really control flow. I don't really understand what you are asking me to do here: to add or not it as cause?

Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/TransactionIdTracking.java]: Add a method `Predicates.await()` that returns a boolean instead of throwing `TimeoutException`. Then check the returned value here and throw other exception with the correct message.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 9 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: I disagree that we need the big refactor for this. You are welcome to try it yourself.

Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: Do you agree that listeners make code more convoluted?

Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: I am not convinced that in this case the code is really convoluted. Moreover I have simply extended a design already in place.

Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: I dislike existing design as well and think that we should not extend it. Here we use callbacks in **single-threaded** code. This is unreasonable.

File: pom.xml (hunk 2 above)
Comment (davidegrohmann) [inline pom.xml]: Well actually it makes some tests to fail, I'll drop this change.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 9 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: Fine, if you dislike the original design, we can change it. But I don't think this PR should be blamed for the current suboptimal design and also I don't think this PR is the right place for that refactor.

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java [hunk 11]
```diff
@@ -704,11 +704,11 @@ RecordStream run( SessionStateMachine ctx, String statement, Map<String, Object>
...
         void udcRegisterClient( String clientName );
         Statement currentStatement();
-        TransactionIdTracker versionTracking( long startingVersion );
+        VersionTracker versionTracker( long startingVersion );
 
-        interface TransactionIdTracker
+        interface VersionTracker
         {
-            void assertUpToDate() throws TransactionFailureException;
+            void assertUpToDate() throws InterruptedException, TransactionFailureException;
```
Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java]: Why `InterruptedException`?

File: community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/StandardStateMachineSPI.java [hunk 12]
```diff
@@ -90,13 +91,23 @@ public void reportError( String message, Throwable cause )
...
             throws TransactionFailureException
     {
-        transactionIdTracker.assertUpToDate();
+        try
+        {
+            versionTracker.assertUpToDate();
+        }
+        catch ( InterruptedException e )
+        {
+            Thread.interrupted();
```
Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/StandardStateMachineSPI.java]: Clear the interruption status? Why?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 9 above)
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: When we discussed this change we've reached a conclusion that `CloseListener` is not the best foundation to build upon. Now this PR even pushes the listeners thing forward. If we do not take a step back now to rethink the design, we'll find ourselves in a situation when proposed refactoring is really hard to do because too many things rely on it.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 10 above)
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java]: `-1` should be a constant in the `CloseListener` than and it's name should be descriptive. It should also be used everywhere instead of `-1` including tests.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 9 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: We agreed to implement  those changes here only if the refactor was trivial. Since it is not, I am not willing to do it as part of this PR.

Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: Refactoring is trivial. I've made it here: https://github.com/neo4j/neo4j/pull/7526.
Please look through the changes there and make use of them in this PR.

Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:147]: I have added some comments there. If you decide to merge that in, I'll rebase this on top of that one with the needed changes.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java (hunk 10 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java]: Will do.

**Comment by davidegrohmann** [comment]
@lutovich I am merging this as is. I'll refactor the way we retrieve txIds when closing transactions in a separate PR.
